General periodic hybrid-DFTB formalism

vanderhe opened the pull request:
General, periodic hybrid-DFTB formalism, as well as a variety of related improvements.

Main features (+ a lot of minor improvements):
- generalization of hybrid-DFTB ground-state Hamiltonian to general CAM functionals
(non-periodic systems)
- MPI-parallelization of matrix-based hybrid-DFTB ground-state Hamiltonian construction
(non-periodic systems)
- introduction of highly optimized matrix-based energy gradient evaluation
(non-periodic systems)
- neighbour-list based hybrid-DFTB ground-state Hamiltonian and energy gradients for general CAM functionals
(periodic, Gamma-only, MPI-parallel version restricted to Hamiltonian construction)
- matrix-based hybrid-DFTB ground-state Hamiltonian and energy gradients for general CAM functionals
(periodic, Gamma-only, MPI-parallel versions available)
- neighbour-list based hybrid-DFTB ground-state Hamiltonian and energy gradients for general CAM functionals
(periodic, general k-points, MPI-parallel version restricted to Hamiltonian construction)
- related infrastructure, including but not limited to new (parallel) dense Mulliken analysis, q0 subtraction, real <--> dual space density matrix transformation, ...

Known issues/missing:
- periodic systems limited to spin-restricted calculations
(routines do support co-linear spin-polarization throughout but haven't been extensively tested yet)
- licensing of Wigner-Seitz related routine needs discussion
- energy gradients of general k-point implementation not yet optimized/MPI-parallelized
- neighbour-list based, Gamma-only Hamiltonian construction could probably be more efficient
- possible incompatibility with other extension due to generalization to general CAM functionals
(hybrid TD-DFTB, ...)
- due to the lack of database, a lot of SK-files checked in for regression testing
- general k-point implementation does not yet utilize MPI-windows and stores Hamiltonian/density matrices on all MPI ranks, leading to relatively high memory consumption
- limited support/testing of restarting calculations
- limitations in k-point sampling, needs to be MP-like with zero shift
- TCI cutoff search might be inefficient, however not performance critical in the slightest
- stress tensor not implemented for all available periodic, hybrid-DFTB ground-state calculations
- slightly inconsistent variable naming in hybridXc module
- general k-point implementation does not yet provide energy gradients for the MIC algorithm
- possible duplicate of ordinary gamma function by full-range HF gamma
- ...

Possible reviewing strategy:
Since this PR includes a lot of changes and added code, reviewing in chunks might be a viable strategy.

1. hybrid-DFTB: rangeseparated.F90, rshgamma.F90, wigner.F90, ...
2. physics related infrastructure: populations.F90, densitymatrix.F90, ...
3. technical infrastructure, parallelization techniques, restarting, initialization, parsing, ...
4. Regression testing
5. Documentation (manual)

bhourahine commented:
@vanderhe Can you please rebase on main when possible (looks like mostly collisions with #1221 )

vanderhe commented:
> @vanderhe Can you please rebase on main when possible (looks like mostly collisions with #1221 )

Sure, I ended up implementing the SK-check a bit differently, compared to this version. I will resolve the conflicts soon.

bhourahine commented:
@vanderhe Just got the buildbot results back, MPI + 2 OMP threads fails several tests (1 OMP passes, so there is something up with at least one openMP construct).

bhourahine commented:
@vanderhe actually, that might just be the lack of UseOmpThreads = Yes in the input files or a block on running these tests with OMP enabled.

vanderhe commented:
> @vanderhe Just got the buildbot results back, MPI + 2 OMP threads fails several tests (1 OMP passes, so there is something up with at least one openMP construct).

I will have a look at it. I used to actively deactivate OMP at the configuration step, since there was no benefit in hybrid parallelization.

bhourahine commented on src/dftbp/dftbplus/main.F90:
```diff
@@ -2881,27 +3088,32 @@ subroutine buildAndDiagDenseRealHam(env, denseDesc, ints, species, neighbourList
-    !> Change in density matrix during last rangesep SCC cycle
-    real(dp), pointer, intent(in) :: deltaRhoInSqr(:,:,:)
+    !> Change in density matrix during last hybridXc SCC cycle
+    real(dp), intent(in) :: deltaRhoIn(:,:,:)
```
As this won't be allocated for non-hybrid calculations
```suggestion
    real(dp), intent(in), allocatable :: deltaRhoIn(:,:,:)
```

vanderhe replied:
Thanks for pointing this out; fixed by now.

vanderhe commented:
> @vanderhe actually, that might just be the lack of UseOmpThreads = Yes in the input files or a block on running these tests with OMP enabled.

@bhourahine Should be fixed, I triggered all buildbot tests again.

vanderhe commented:
@bhourahine Dangling pointers/missing allocatables detected by NAG fixed.

bhourahine commented:
@vanderhe the scalapackfx version seems to be behind the one main is using.

vanderhe commented:
> @vanderhe the scalapackfx version seems to be behind the one main is using.

Yes, let me fix that real quick.

vanderhe commented:
@bhourahine The scalapackfx tag should be correct now but how do I handle the [submodules](https://github.com/dftbplus/dftbplus/pull/1223/files#diff-d6c28fb5dbf4b3b3cdb4df36d00feff524b5df4916822cda3ddaa8945affd8c2)?

vanderhe commented:
> @vanderhe the scalapackfx version seems to be behind the one main is using.

@bhourahine Should be fixed by now, had to remove some changes to the submodules.

bhourahine commented on src/dftbp/math/wigner.F90:
```diff
@@ -0,0 +1,251 @@
+      call error('Wigner-Seitz Module: Weights do not add up to nk(1) * nk(2) * nk(3)')
+    end if
+
+    !@ JN it happens in 2d and 1d systems with small course grids. I've changed to 20**3
```
!@ and JN ???

bhourahine commented on src/dftbp/timedep/timeprop.F90:
```diff
@@ -3213,10 +3210,15 @@ subroutine getForces(this, movedAccel, totalForce, rho, H1, Sinv, neighbourList,
-          & nNeighbourSK)
+    if (this%isHybridXc) then
+    #:if WITH_SCALAPACK
+      call error("Timeprop: MPI-parallel hybrid-DFTB matrix-based force evaluation not&
```
Pass upwards, as with other error handling

vanderhe replied:
done

bhourahine commented on src/dftbp/timedep/timeprop.F90:
```diff
@@ -51,7 +53,7 @@ module dftbp_timedep_timeprop
   use dftbp_elecsolvers_elecsolvers, only : TElectronicSolver
   use dftbp_extlibs_tblite, only : TTBLite
-  use dftbp_io_message, only : warning
+  use dftbp_io_message, only : error, warning
```
revert

bhourahine commented on src/dftbp/dftb/sccinit.F90:
What happens if this is given an earlier version of a restart file to read? Can if fill something in for the DM or fail gracefully?

vanderhe replied:
It now throws an error, should the restart file version undercut the earliest format that supported general (periodic) hybrids.

bhourahine commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4426,38 +4426,46 @@ \subsubsection{DFTB3-D3H5}
+parametrization was not created for the respective hybrid xc-functional,
+\dftbp{} will issue an error message and halt.
+
+The \kw{Hybrid} keyword expects either \is{None} (default -- no use of hybrid functional),
```
Definitely need an explantion of the different functionals (possibly in the table suggested elsewhere for the manual).

vanderhe replied:
I have extended the corresponding manual entry and tried to guide the user, depending on the boundary conditions and further requirements (MPI, colinear spin, restart capability, etc.).

bhourahine commented on src/dftbp/dftb/rshgamma.F90:
```diff
@@ -0,0 +1,500 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2022  DFTB+ developers group                                               !
```
```suggestion
!  Copyright (C) 2006 - 2023  DFTB+ developers group                                               !
```

bhourahine commented on src/dftbp/dftb/rshgamma.F90:
```diff
@@ -0,0 +1,500 @@
+          & / ((endDiag1(2) - startDiag1(2)) * (startDiag2(1) - endDiag2(1))&
+          & - (endDiag1(1) - startDiag1(1)) * (startDiag2(2) - endDiag2(2)))
+
+      ! check if third eq. also satisfied (3 equations but 2 variables)
```
thrid eq. of what? Add paper ref somewhere in this source.

vanderhe replied:
This is simple trigonometry really, I think I even over-complicated this implementation. One equation for each component (x,y,z). I will slightly adapt the comment(s).

bhourahine commented on src/dftbp/dftb/rshgamma.F90:
```diff
@@ -0,0 +1,500 @@
+        tau = 0.5_dp * (tauA + tauB)
+        gamma = tau * 0.3125_dp
+      else
+        call error("Error(RSH-Gamma): R = 0, Ua != Ub")
```
Pass errors upwards.

vanderhe replied:
I made the functions "error propagation ready", someone else with a better understanding of it can then pass it upwards as desired.

bhourahine commented on src/dftbp/dftb/rshgamma.F90:
```diff
@@ -0,0 +1,500 @@
+
+  use dftbp_common_accuracy, only : dp, tolSameDist, MinHubDiff
+  use dftbp_math_simplealgebra, only : cross3
+  use dftbp_io_message, only : error
```
Use the type and macro for error handing.

bhourahine commented on src/dftbp/dftbplus/main.F90:
```diff
@@ -767,43 +782,100 @@ subroutine processOutputCharges(env, this)
-      case default
-        call error("Range separation not implemented for non-colinear spin")
-      end select
+    ! CAM calculations deduct atomic charges from delta density matrix
```
```suggestion
    ! CAM calculations need to deduct atomic charges from delta density matrix
```

bhourahine commented on src/dftbp/dftbplus/main.F90:
```diff
@@ -767,43 +782,100 @@ subroutine processOutputCharges(env, this)
     type(TDftbPlusMain), intent(inout) :: this
 
-    integer :: iSpin
+    ! Square dense overlap storage for calculating delta Rho in Gamma-point case with
```
```suggestion
    ! Square dense overlap storage for calculating delta rho in Gamma-point case with
```

bhourahine commented on src/dftbp/dftbplus/main.F90:
```diff
@@ -2878,27 +3067,32 @@ subroutine buildAndDiagDenseRealHam(env, denseDesc, ints, species, neighbourList
 
     !> Coordinates of all atoms including images
-    real(dp), allocatable, intent(inout) :: coord(:,:)
+    real(dp), intent(inout), allocatable :: coord(:,:)
```
???

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
-  end type TRangeSepTypesEnum
+  type :: TIntArray1D
+
+    !> Onedimensional, integer data storage
```
```suggestion
    !> One dimensional integer data storage
```

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+
+  type :: TRealArray1D
+
+    !> Onedimensional, real data storage
```
```suggestion
    !> One dimensional real data storage
```

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+
+  type :: TRealArray2D
+
+    !> Twodimensional, real data storage
```
```suggestion
    !> Two dimensional real data storage
```

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
   !> Enumerator for type of hybrid functional used.
-  !! (limited to purely long-range corrected for now)
-  type :: TRangeSepFuncEnum
+  !! (global hybrid, purely long-range corrected, general CAM range-separated form)
```
Misses none

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+        end if
+      end do
+    end do lpOuter
+    if (tNotMonkhorstPack) then
```
What about k-lines? I.e. band structures.

vanderhe replied:
See above.

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+    ! range-separated type is initialized with nAtom0 coordinates, therefore re-allocate for
+    ! periodic systems, where images beyond the central cell are accounted for
+    if (allocated(this%coords)) deallocate(this%coords)
+    if (allocated(this%rCoords)) deallocate(this%rCoords)
+    this%rCoords = symNeighbourList%coord
+    allocate(this%coords(size(this%rCoords, dim=1), size(this%rCoords, dim=2)))
+
+    ! Calculate neighbour list coordinates in relative units
+    this%coords(:,:) = this%rCoords
+    call cart2frac(this%coords, latVecs)
+
+    ! re-allocate sparse overlap for symmetric neighbour list
+    if (allocated(this%overSym)) deallocate(this%overSym)
+    allocate(this%overSym(symNeighbourList%sparseSize))
 
-    nOrb = size(overlap,dim=1)
+    ! build symmetric, sparse overlap
+    call buildS(env, this%overSym, skOverCont, this%rCoords, nNeighbourCamSym,&
+        & symNeighbourList%neighbourList%iNeighbour, symNeighbourList%species,&
+        & symNeighbourList%iPair, orb)
 
-    allocate(Smat(nOrb,nOrb))
-    allocate(Dmat(nOrb,nOrb))
-    allocate(LrGammaAO(nOrb,nOrb))
-    allocate(gammaCmplx(nOrb,nOrb))
-    allocate(Hlr(nOrb,nOrb))
+    if (this%tScreeningInited) then
+      this%hPrevCplxHS(:,:,:) = 0.0_dp
+      this%dRhoPrevCplxHS(:,:,:,:,:,:) = 0.0_dp
+      this%camEnergy = 0.0_dp
+    end if
+
+    ! pre-tabulate overlap estimates for neighbour-list based algorithms
+    ! if-branch not really necessary, since k-implementation only available for neighbour-list
+    ! based algorithm anyway, but let's be on the save side
+    if (this%hybridXcAlg == hybridXcAlgo%neighbourBased) then
+      call calculateOverlapEstimates(this, symNeighbourList, nNeighbourCamSym, iSquare)
+    end if
+
+    ! ##################### Beginning of \gamma(\bm{g}) pre-tabulation #############################
```
Comment style???

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+
+  !> Returns the numerical second derivative of long-range gamma, by means of a central finite
+  !! difference.
+  function getddLrNumericalGammaValue(this, iSp1, iSp2, dist, delta) result(ddGamma)
```
Should differentiate this analytically at some point...

vanderhe replied:
yes

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -8,1627 +8,6137 @@
+
+  end subroutine tabulateCamdGammaEval0_gamma
+
+
```
```suggestion
```

bhourahine reviewed (commented):
- Buildbot fails for timedep/Furan-NACV with several compilers and md/Vsi+O-plumed for one case.
- Check copyright years on new files (think I found most of them).
- parser version will have to be 14 (and date introduced changed)
- various extraneous edits (white space changes mostly), which might as well stay, but should have been other PRs to speed up reviewing.
- Files need proper line ending on last line (think I found most of them).
- Don't use error(), pass errors upwards to the highest level of the code with the relevant type and macros (so they can eventually be properly handled for API calls, instead of bringing the code down uncontrollably).
- Make the instance of the symmetric neighbour structure allocatable
- Why is the new overlap required over the BvK index, instead of building dynamically? It's quite a large matrix.
- Can you document various data structures somewhere, but also in the manual give some more information about algorithm choice and functional nature/compatibilty?

I'll need to go over rangeseparated.F90 again more carefully.

bhourahine reviewed (commented):
- Buildbot fails for timedep/Furan-NACV with several compilers and md/Vsi+O-plumed for one case.
- Check copyright years on new files (think I found most of them).
- parser version will have to be 14 (and date introduced changed)
- various extraneous edits (white space changes mostly), which might as well stay, but should have been other PRs to speed up reviewing.
- Files need proper line ending on last line (think I found most of them).
- Don't use error(), pass errors upwards to the highest level of the code with the relevant type and macros (so they can eventually be properly handled for API calls, instead of bringing the code down uncontrollably).
- Make the instance of the symmetric neighbour structure allocatable
- Why is the new overlap required over the BvK index, instead of building dynamically? It's quite a large matrix.
- Can you document various data structures somewhere, but also in the manual give some more information about algorithm choice and functional nature/compatibilty?

I'll need to go over rangeseparated.F90 again more carefully.

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4426,38 +4426,46 @@ \subsubsection{DFTB3-D3H5}
+For the non-periodic formalism we refer to
+Refs.~\cite{niehaus-PSSB-249-237,lutsker-JCP-143-184107}, whereas the formalism for
+periodic boundary conditions is developed in Ref.~\cite{vdH-JCP-???-??????}.
+The electron-electron interaction is partitioned into short- and long-range contributions
```
done

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
       cutoff, beyond which the overlap between atoms is considered to be zero.
-      This can be used as an additional tweak to speed up the LC-calculation,
+      This can be used as an additional tweak to speed up the hybrid-calculation,
       but make sure first, that your results do not change
```
done

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
+      periodicity of the density matrix for finite k-point samplings.
+    \item[\is{GSummationCutoff}]\modif{\modtype{length}} Spatial cutoff of g-summations
+      of Ref.~\cite{}. Usually a modification of the default value by the user is not
+      necessary. Tweak only if you know what you are doing.
```
done

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
+      necessary. Tweak only if you know what you are doing.
+    \end{description}
+    
+    Note: Recommended choice for calculations where analytic energy gradients are required.
```
E.g. geometry relaxation, MD, ... now added explicitly. I also remove "analytic".

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
+      \is{Truncated} block.
+    \end{description}
+    
+  \item[\is{MIC \cb}] Adaption of the minimum image convention to avoid divergent lattice
```
done

vanderhe commented on src/dftbp/dftbplus/parser.F90:
```diff
@@ -7713,108 +7733,184 @@ function is_numeric(string) result(is)
+        select case(char(buffer))
+        case ("full")
+          input%gammaType = hybridXcGammaTypes%full
+        case ("mic")
```
done

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
 
-Example for neighbour list based screening with customised cutoff reduction:
+Example for neighbour-list based screening with customised cutoff reduction
+(purely long-range corrected functional):
```
The screening choice does not depend on the type of hybrid functional used, only the boundary conditions, parallelism, spin-formalism, etc.
I have now indicated that the choice of functional is also just an example.

vanderhe commented on doc/dftb+/manual/dftbp.tex:
```diff
@@ -4469,55 +4477,183 @@ \subsection{RangeSeparated}
 \end{verbatim}
 
-Example for matrix-matrix multiplication based method:
+Example for matrix-matrix multiplication based method (purely long-range
```
See above.

vanderhe commented on src/dftbp/common/parallel.F90:
```diff
@@ -0,0 +1,58 @@
+contains
+
+  !> Returns the start and end index of an MPI process that calculates parts of a loop.
+  pure subroutine getStartAndEndIndex(nSystems, nProcs, iProc, iStart, iEnd)
```
Could be the case, I will try to make use of it and remove the duplicate.

vanderhe commented on src/dftbp/dftb/densitymatrix.F90:
```diff
@@ -15,37 +15,124 @@
   end interface makeDensityMatrix
 
+
+  !> Holds real and complex delta density matrices and pointers.
```
Also, does not encompass pointers anymore, done.

vanderhe commented on src/dftbp/dftb/densitymatrix.F90:
```diff
@@ -15,37 +15,124 @@
+    !> Composite index for mapping iK/iS --> iGlobalKS for arrays present at every MPI rank
+    integer, allocatable :: iKiSToiGlobalKS(:,:)
+
+  ! #:if WITH_SCALAPACK
```
No, debugging artifact, should be removed.

vanderhe commented on src/dftbp/dftbplus/oldcompat.F90:
```diff
@@ -829,6 +828,13 @@ subroutine convert_12_13(root)
 
     end if
 
+    call getDescendant(root, "Hamiltonian/DFTB/Rangeseparated", ch1)
```
Yes, done.

vanderhe commented on src/dftbp/extlibs/poisson.F90:
```diff
@@ -24,6 +24,9 @@ module dftbp_extlibs_poisson
   use libmpifx_module, only : mpifx_barrier, mpifx_bcast
 #:endif
 #:if WITH_POISSON
+#:if WITH_MPI
```
Good question, I cannot remember having changed this file on purpose (could accidentally have happened due to #1118). This is just a duplicate of the imports below, so I will remove it.

vanderhe commented on test/app/dftb+/hybrid/periodic/GaAs-kpts-mic/dftb_in.hsd:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  KPointsAndWeights = SupercellFolding {
```
Shifts not yet supported and I remember that it was somehow relevant to go beyond 2x2x2 for testing translational invariance (I think...).

vanderhe commented on test/app/dftb+/hybrid/cluster/C18H12-force-cam/dftb_in.hsd:
```diff
@@ -0,0 +1,44 @@
+}
+
+ParserOptions = {
+  ParserVersion = 5
```
I kept it on purpose, just to test backwards compatibility.

vanderhe commented on test/app/dftb+/hybrid/cluster/C18H12-matrix/dftb_in.hsd:
```diff
@@ -38,17 +37,18 @@ Driver = {}
 Hamiltonian = DFTB {
   SCC = Yes
-  SCCTolerance = 1.0E-8  # Extremely small
+  SCCTolerance = 1.0E-10  # Extremely small
```
I think this was related to something about the forces.

vanderhe commented on test/app/dftb+/hybrid/cluster/C18H12-matrix/dftb_in.hsd:
```diff
@@ -38,17 +37,18 @@ Driver = {}
-    C-H = "C-H.skf"
-    H-C = "H-C.skf"
-    H-H = "H-H.skf"
+  Mixer = Simple {MixingParameter = 0.85}
```
Using the simple mixer made testing the restart functionality much easier, this is why I used it occasionally, but I have to think about this particular change, not sure to be honest.

vanderhe commented on test/app/dftb+/hybrid/cluster/C60-nb-cutoff/dftb_in2.hsd:
```diff
@@ -10,8 +10,10 @@ Hamiltonian = DFTB {
 
-    SlaterKosterFiles = {
-        C-C = "C-C.skf"
+    SlaterKosterFiles = Type2FileNames {
```
I just wanted to get rid of as many symlinks as possible.

vanderhe commented on test/app/dftb+/hybrid/periodic/Acene-gamma-matrix-force-cam/dftb_in.hsd:
```diff
@@ -0,0 +1,61 @@
+  Hybrid = CAM {
+    CoulombMatrix = Truncated {
+      CoulombCutoff = 2.885010
+      GSummationCutoff = 50.985000
```
done

vanderhe commented on test/app/dftb+/hybrid/periodic/Acene-gamma-matrix-force-cam/dftb_in.hsd:
```diff
@@ -0,0 +1,61 @@
+    Screening = MatrixBased {}
+  }
+
+  KPointsAndWeights = SupercellFolding {
```
I blocked everything except for the supercell folding technique, so this would actually result in a (useful) error message. Not nice thought, since both ways are in principle equivalent, I just spared some additional logic by that choice.

vanderhe replied:
I have now added some additional logic to accept any k-point specification that leads to a Gamma-only sampling.

vanderhe commented on test/app/dftb+/hybrid/periodic/Benzene-gamma-neigh-spin-force/dftb_in.hsd:
```diff
@@ -0,0 +1,60 @@
+  Hybrid = LC {
+    CoulombMatrix = Truncated {}
+    Screening = NeighbourBased {
+      Threshold = 1e-100
```
Just wanted to make sure that there is no influence whatsoever of the exact implementation of the pre-screening (as this is most likely subjected to changes in the future).

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -1137,8 +1160,8 @@ module dftbp_dftbplus_initprogram
-    procedure :: ensureRangeSeparatedReqs
-    procedure :: initRangeSeparated
+    procedure :: ensureHybridXcReqs
+    procedure :: reallocateHybridXc
```
Changed as well.

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -891,6 +911,9 @@ module dftbp_dftbplus_initprogram
     !> Square dense overlap storage for cases with k-points
     complex(dp), allocatable :: SSqrCplx(:,:)
 
+    !> Square dense overlap that actually stores all k-points
```
At the moment it is needed for q0 subtraction but changing it to "unpack-when-needed" should be straight-forward and not performance critical.

vanderhe replied:
Changed to "unpack-when-needed" strategy.

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -265,16 +286,20 @@ module dftbp_dftbplus_initprogram
     type(TNeighbourList), allocatable :: neighbourList
 
+    !> ADT for neighbour parameters, symmetric version for CAM calculations
+    type(TSymNeighbourList) :: symNeighbourList
```
Just a historical artifact, we can change it I guess.

vanderhe replied:
done

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -1699,11 +1729,17 @@ subroutine initProgramVariables(this, input, env)
     allocate(this%img2CentCell(this%nAllAtom))
     allocate(this%iCellVec(this%nAllAtom))
 
+    ! Copy symmetry-specific entries into symmetric neighbour list
```
Yeah, the comment was outdated and the allocation should not happen for non-hybrid runs, done.

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -3409,28 +3541,65 @@ subroutine initProgramVariables(this, input, env)
+    if (this%isHybridXc) then
+      if (input%ctrl%hybridXcInp%hybridXcType == hybridXcFunc%hyb) then
+        write(stdOut, "(A,':',T30,A)") "Global hybrid", "Yes"
+        write(stdOut, "(2X,A,':',T30,E14.6)") "Fraction of HF",&
```
I know, but I thought this is what the users are more familiar with. Changed to "exchange" nevertheless.

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -3894,7 +4063,7 @@ end subroutine setEquivalencyRelations
 
     !> Instance
-    class(TDftbPlusMain), intent(inout) :: this
+    class(TDftbPlusMain), intent(inout), target :: this
```
Artifact of a previous version that was relying on pointers, I have removed the target attribute.

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -5868,106 +6236,105 @@ subroutine printReksInitInfo(reks, orb, speciesName, nType)
-          write (stdOut, "(A,':',T30,A)") "CP-REKS Solver", "Conjugate-Gradient"
-          write (stdOut, "(A,':',T30,I14)") "CG max. Iterations", reks%CGmaxIter
-          write (stdOut, "(A,':',T30,E14.6)") "CG Tolerance", reks%Glimit
+        elseif (reks%Glevel == 2) then
```
Unrelated changes, sorry for that...

vanderhe commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -6026,7 +6393,7 @@ subroutine initGeometry_(input, nAtom, nType, tPeriodic, tHelical, boundaryCond,
       cellVol = abs(determinant33(latVec))
       recCellVol = abs(determinant33(recVec))
-    else if (tHelical) then
+    elseif (tHelical) then
```
Again, unrelated changes, sorry. What is actually our convention for else-if statements, or is it a 50:50 thing?

vanderhe commented on src/dftbp/dftb/populations.F90:
```diff
@@ -487,28 +926,138 @@ subroutine denseSubtractDensityOfAtoms_spin_cmplx(q0, iSquare, rho, iSpin)
+
+    nAtom = size(q0, dim=2)
+
+    ! distinguish cases: spin-unpolarized, colinear-spin
```
done

vanderhe commented on src/dftbp/dftb/sparse2dense.F90:
```diff
@@ -817,7 +818,7 @@ subroutine packHS_cmplx_kpts(primitive, square, kPoint, kWeight, iNeighbour, nNe
     @:ASSERT(all(shape(kPoint) == [3]))
     @:ASSERT(all(shape(nNeighbourSK) == [nAtom]))
-    @:ASSERT(kWeight > 0.0_dp)
+    ! @:ASSERT(kWeight > 0.0_dp)
```
Artifact of fake k-point procedure that I tried a year ago or so, commented in again.

vanderhe commented on src/dftbp/dftb/sparse2dense.F90:
```diff
@@ -2021,6 +2022,21 @@ subroutine hermitianSquareMatrix(matrix)
   end subroutine hermitianSquareMatrix
 
 
+  !> copy lower triangle to upper for a square matrix
```
I removed the whole routine, since it is not used anywhere.

vanderhe commented on src/dftbp/dftbplus/input/geoopt.F90:
```diff
@@ -14,14 +14,14 @@ module dftbp_dftbplus_input_geoopt
 
   private
-  public :: readGeoOptInput, TGeoOptInput
+  public :: readGeoOptInput, readOptimizerInput, TGeoOptInput
```
Added in advance, because I already knew it is later used by the electronic constraints, so not useless anymore because cDFTB has already been merged ;)

vanderhe commented on src/dftbp/math/wigner.F90:
I still have to rewrite this whole file...

vanderhe replied:
Renamed to wignerseitz.F90.

vanderhe commented:
> I'm getting major fails on buildbot with the current code state: forrtl: severe (408): fort: (7): Attempt to use pointer SYMNEIGHBOURLIST when it is not associated with a target dftb+ 000000000333708F Unknown Unknown Unknown dftb+ 0000000000A50C89 dftbp_dftbplus_ma 1315 main.f90 dftb+ 0000000000A012B2 dftbp_dftbplus_ma 214 main.f90 dftb+ 000000000042EDA5 MAIN__ 32 dftbplus.f90 Probably symNeighbourList should be allocatable in getDensity() in main.F90

Yes, I think I forgot to properly propagate the allocatable attribute downwards, I'm still working on some of your (non-trivial) feedback so changes might occur frequently.

vanderhe commented on test/app/dftb+/hybrid/periodic/Acene-kpts-restart/dftb_in2.hsd:
```diff
@@ -0,0 +1,66 @@
+Hamiltonian = DFTB {
+
+  SCC = Yes
+  SccTolerance = 1e-06
```
I have revised all regression tests added, that cover restart functionality, to your proposed strategy.

vanderhe commented:
@bhourahine I think I have addressed all your remarks (except for the Wigner-Seitz issue). Ready for the next iteration, feel free to have a look at it again.

bhourahine commented:
@vanderhe The Furan problem case is due to the deltaRho passed to addNadiaGradients() in linrespgrad.F90 being an optional in the calling routine, but not optional inside the routine.

vanderhe commented:
> @vanderhe The Furan problem case is due to the deltaRho passed to addNadiaGradients() in linrespgrad.F90 being an optional in the calling routine, but not optional inside the routine.

@bhourahine Thanks Ben, now I understand why I did not observe it locally. I was only testing timedep/Furan-NACV-LC, not timedep/Furan-NACV, since I usually do not compile with ARPACK. Shouldn't timedep/Furan-NACV be executed even if ARPACK is disabled?

bhourahine commented on src/dftbp/timedep/linrespgrad.F90:
```diff
@@ -133,7 +133,7 @@ subroutine LinRespGrad_old(this, iAtomStart, grndEigVecs, grndEigVal, sccCalc, d
 
     !> difference density matrix (vs. uncharged atoms)
-    real(dp), intent(inout), pointer :: deltaRho(:,:,:)
+    real(dp), intent(inout), optional :: deltaRho(:,:,:)
```
This change is the origin of the Furan fail with intel. Either addnadiagradients should be adapted to also make deltaRho optional, or both need to change to it being allocatable.

vanderhe replied:
See last commit, but this is not robust. At some point we should check if all optional dummy arguments are actually present before just blindly using them. This concerns multiple places in linrespgrad I guess.

vanderhe replied:
I triggered all tests again, see: `Builders/trigger_all_tests/43`

bhourahine commented:
> @bhourahine Thanks Ben, now I understand why I did not observe it locally. I was only testing timedep/Furan-NACV-LC, not timedep/Furan-NACV, since I usually do not compile with ARPACK. Shouldn't timedep/Furan-NACV be executed even if ARPACK is disabled?

The non-LC test is using the DFTB2 model and the older ARPACK solver.

bhourahine commented:
@vanderhe There is still the problem with md/Vsi+O-plumed (see the recent buildbot tests). It looks like uninitialised variables somewhere, possibly somewhere in the reallocation of the sparse matrices.

vanderhe commented:
> @vanderhe There is still the problem with md/Vsi+O-plumed (see the recent buildbot tests). It looks like uninitialised variables somewhere, possibly somewhere in the reallocation of the sparse matrices.

@bhourahine Thanks for the hint, I only had a brief look at it and posponed working on it as I saw that it isn't absolutely trivial to fix, but rather exhibits a strange behavior in the SCC cycles that suddenly explode and NaN entries appear. I will try to find time for it this week.

bhourahine commented:
@vanderhe I think what is happening is that after the SCC cycle there is some variable used unititialised, where since the compiler is setting these to NaNs by default it propagates and explodes. This could be anything from reallocation of various arrays on geometry change without initialising to 0, throught to the wrong intent for a variable (out vs inout). I don't think it is a numerical fail.

vanderhe commented:
> @vanderhe I think what is happening is that after the SCC cycle there is some variable used unititialised, where since the compiler is setting these to NaNs by default it propagates and explodes. This could be anything from reallocation of various arrays on geometry change without initialising to 0, throught to the wrong intent for a variable (out vs inout). I don't think it is a numerical fail.

@bhourahine The NAG and Intel22 compilers take the same SCC path and the Hamiltonian and overlap matrices of the last SCC step  (that then fails for NAG during the diagonalization) are virtually identical, see the tagdiff report of me writing out those quantities (`MaxSCCIterations = 9`):
```
hamiltonian          element              2.0261570199409107e-15      OK        
overlap              element              0.0                         OK
```
I have some doubts that the issue is really on our side.

vanderhe commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -6,1629 +6,6250 @@
-    !> Threshold based
-    integer :: threshold = 1
+  !> Returns the (directed) derivative of long-range + full-range Hartree-Fock gamma.
+  interface getDirectedCamGammaPrimeValue
```
It is just $\hat{\vec{v}}\frac{\partial \gamma(R)}{\partial R}$, with $\hat{\vec{v}}$ being the unit vector pointing from atom 2 to atom 1. How shall we call this thing?

bhourahine replied:
That is the directional derivative of γ

vanderhe replied:
I see, that means `directed --> directional` in various places? Will do so.

vanderhe commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -6,1629 +6,6250 @@
 !--------------------------------------------------------------------------------------------------!
 
 #:include 'common.fypp'
+#:include 'error.fypp'
```
I did not yet do anything about this, you resolved it intentionally?

bhourahine replied:
The new code in getddHfNumericalGammaDeriv doesn't match the intention of the macro, as it uses a local error() call inside the function, instead of passing the messenger variable upwards (and also calls the propogate macro, which is broken as this will just return the function in this case). These message variables are supposed to intent(out) all of the way up to the level of main.F90, where it then gets delt with.

bhourahine replied:
> I did not yet do anything about this, you resolved it intentionally?
Nope, I didn't do it either.

vanderhe replied:
Hm, my GUI says `bhourahine marked this conversation as resolved.`, seems to be a buggy. Let me unresolve it then.

vanderhe replied:
I had another discussion with Bálint and what is implemented right (in particular rshGamma.F90 and rangeseparated.F90) now essentially reflects the outcome of that.

bhourahine commented on src/dftbp/dftb/rangeseparated.F90:
```diff
@@ -6,1629 +6,6250 @@
+    class(THybridXcFunc), intent(in) :: this
+
+    !> Gradient of gamma between atoms
+    real(dp), intent(out) :: grad(3)
```
Did this one also get resolved erroneously?

vanderhe replied:
This I actually applied in the latest commit [95123fb](https://github.com/dftbplus/dftbplus/pull/1223/commits/95123fbfc7b06bf75f02f74bf3c0a826562c9d52) (2 places).

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+  end subroutine appendToArray1d
+
+
+  !> Appends element to two-dimensional, integer-valued array.
```
```suggestion
  !> Appends a row to a two-dimensional, integer-valued array.
```

vanderhe replied:
Makes sense, will apply soon.

bhourahine commented on src/dftbp/math/wignerseitz.F90:
Any reason not to make appendToArray?d general and put them somewhere accessible by other places in the code? Similar code is already in dftb/periodic.F90/reallocateArrays*() but that implementation also sizes with incrmntOfArray to slightly reduce the number of allocation needed if called multiple times.

vanderhe replied:
I didn't really like these two new routines, precisely because of all the (re-)allocations, that's why I kept them inside this file. At them moment, at least
https://github.com/dftbplus/dftbplus/pull/1223/files#diff-434655a60c1dbfca91c3409342660de33a74204f94ae438b638779ed6f3c846cR125
relies on an increment of 1, not necessarily a good reason not to generalize though.
If you think that these routines might be helpful to others, I can generalize them.

bhourahine replied:
We already have essentially the same code elsewhere, so makes sense to put something general in dftbp_common_memman for integer and real array reallocation with copy, as this is where incrmntOfArray() is currently living.
Might be informative to test how many times appendToArray is called in a typical case here and decide whether or not to use incrmntOfArray() for the sizing.

vanderhe replied:
I have generalized the respective routines to (logical, integer, real, complex) and moved them into a dedicated module append.F90.

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+      distTol_ = 1.0e-08_dp
+    end if
+
+    ! Cache repeatedly used dot-products
```
Isn't that just a GEMM (or even a SYMM) product between latVecs and latVects.T ?

vanderhe replied:
I might have missed that, let me think about it.

vanderhe replied:
So `dotProducts == transpose(dotProducts)` is true, but also `dotProducts /= matmul(latVecs, transpose(latVecs))`, if that is what you meant.

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+      @:ASSERT(size(wsSearchSize, dim=1) == 3)
+      wsSearchSize_(:) = wsSearchSize
+    else
+      wsSearchSize_(:) = 2
```
Is that large enough?

vanderhe replied:
Apparently yes, at least for any somewhat reasonable system. We could bump the default to 3, it should not have a huge impact on performance.

bhourahine replied:
What about pathological cases, what would it take to break this? My guess is having one very long vector.

vanderhe replied:
If you really try you can probably construct such a case, but how do we avoid that?

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+          do i1 = -wsSearchSize_(1), wsSearchSize_(1)
+            do i2 = -wsSearchSize_(2), wsSearchSize_(2)
+              do i3 = -wsSearchSize_(3), wsSearchSize_(3)
+                diff(1) = n1 - i1 * nKpt(1)
```
Compiler optimisation is probably moving this already, but can evaluate after line 112

vanderhe replied:
True, consider it done.

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+            do i2 = -wsSearchSize_(2), wsSearchSize_(2)
+              do i3 = -wsSearchSize_(3), wsSearchSize_(3)
+                diff(1) = n1 - i1 * nKpt(1)
+                diff(2) = n2 - i2 * nKpt(2)
```
Likewise, after line 113

bhourahine commented on src/dftbp/math/wignerseitz.F90:
```diff
@@ -0,0 +1,216 @@
+                diff(3) = n3 - i3 * nKpt(3)
+                do jj = 1, size(latVecs, dim=2)
+                  do ii = 1, size(latVecs, dim=2)
+                    diffMat(ii, jj) = real(diff(ii), dp) * real(diff(jj), dp)
```
Its fine, but could write more concisely as a spread() of the diff vector into a 3x3 matrix and a Hadamard product with the matrix transpose.

vanderhe replied:
What about `diffMat(:,:) = spread(diff, 1, 3) * spread(diff, 2, 3)` to spare the transpose?

vanderhe replied:
The conversion to double-precision reals is carried out automatically in this way, isn't it?

bhourahine commented on src/dftbp/common/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ set(curdir ${CMAKE_CURRENT_SOURCE_DIR})
 
 set(sources-fpp
   ${curdir}/accuracy.F90
+  ${curdir}/append.F90
```
Possibly matrixappend.F90 would be clearer?

vanderhe replied:
I think so, changed by now.

bhourahine reviewed (commented):
After this gets merged, we need a new PR to adapt this structure for the reallocateArrays* calls from dftbp/dftb/periodic.F90 and also to allow for use of incrmntOfArray. The intention is to minimize reallocations and avoid fragmentation of memory while doing it.

bhourahine commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -1713,11 +1742,20 @@ subroutine initProgramVariables(this, input, env)
     allocate(this%img2CentCell(this%nAllAtom))
     allocate(this%iCellVec(this%nAllAtom))
 
+    if (this%isHybridXc) then
```
Probably a good idea to have a general isSymNeighbourNeeded variable, which is set = isHybridXc for the moment, as this would make it easier to generalise other cases later. (will be needed to fix solvation)

vanderhe replied:
I guess this is addressed by your PR for now.

bhourahine commented on src/dftbp/dftb/periodic.F90:
```diff
@@ -94,6 +95,36 @@ module dftbp_dftb_periodic
+    integer, allocatable :: iPair(:,:)
+
+    !> Total number of elements in a sparse structure (ignoring extra indices like spin)
+    integer :: sparseSize
```
Does this really belong in this data structure? It is also related to basis set size, not purely geometry.

vanderhe replied:
No, I think I "temporarily" added it to the sym-neighlist type in order to avoid overwriting the usual `this%sparseSize` variable and forgot to search for a more appropriate place afterwards. What about introducing `this%sparseSizeSym` or something similar?

bhourahine replied:
The cleaner thing might be a type for sparse structure descriptors (we have something sort of like this for the dense case).

bhourahine reviewed (commented):
I'm now getting regression failures for 2 MPI ranks for 
hybrid/cluster/C18H12-force-*

vanderhe commented:
> I'm now getting regression failures for 2 MPI ranks for hybrid/cluster/C18H12-force-*

Passes locally on my machine, which compiler/build configuration is affected?

bhourahine commented:
This was on a local cluster with Intel (21 I think, will check) and 2 procs.

vanderhe commented:
> This was on a local cluster with Intel (21 I think, will check) and 2 procs.

Even worse for me, getting an internal compiler error for rangeseparated.f90...

env_dftb+/intel2021-impi
cmake -DWITH_OMP=1 -DWITH_MPI=1 -DTEST_MPI_PROCS=2 -DCMAKE_BUILD_TYPE=Debug -B _build .

bhourahine reviewed (approved):
Ready to go on CI completion.
